File: remocra/src/main/webapp/js/app/remocra/model/OldebVisite.js
Ext.require('Sdis.Remocra.network.RemocraRest');
Ext.require('Sdis.Remocra.model.OldebVisiteSuite');
Ext.require('Sdis.Remocra.model.TypeOldebAnomalie');
Ext.define('Sdis.Remocra.model.OldebVisite', {
    extend: 'Ext.data.Model',

    fields: [{
        name: 'id',
        type: 'int',
        useNull: true
    }, {
        name: 'code',
        type: 'string'
    }, {
        name: 'dateVisite',
        type: 'date',
        dateFormat: 'c'
    }, {
        name: 'agent',
        type: 'string'
    }, {
        name: 'observation',
        type: 'string'
    }, {
        name: 'nomAvis',
        type: 'string'
    }, {
        name: 'nomAction',
        type: 'string'
    }, {
        name: 'nomDebAcces',
        type: 'string'
    }, {
        name: 'nomDebParcelle',
        type: 'string'
    }, {
        name: 'totalAnomalies',
        type: 'string'
    }, {
        name: 'oldeb',
        type: 'fk',
        useNull: true
    }, {
        name: 'debroussaillementParcelle',
        type: 'fk',
        useNull: true
    }, {
        name: 'debroussaillementAcces',
        type: 'fk',
        useNull: true
    }, {
        name: 'avis',
        type: 'fk',
        useNull: true
    }, {
        name: 'action',
        type: 'fk',
        useNull: true
    } ],

    associations: [{
        type: 'hasMany',
        model: 'Sdis.Remocra.model.OldebVisiteDocument',
        associationKey: 'oldebVisiteDocuments',
        name: 'oldebVisiteDocuments',
        persist: false
    }, {
        type: 'hasMany',
        model: 'Sdis.Remocra.model.TypeOldebAnomalie',
        associationKey: 'typeOldebAnomalies',
        name: 'typeOldebAnomalies',
        persist: true
    }, {
        type: 'hasMany',
        model: 'Sdis.Remocra.model.OldebVisiteSuite',
        associationKey: 'oldebVisiteSuites',
        getterName: 'getOldebVisitesSuites',
        setterName: 'setOldebVisitesSuites',
        name: 'oldebVisiteSuites',
        persist: true
    } ],

    proxy: {
        type: 'remocra.rest',
        url: Sdis.Remocra.util.Util.withBaseUrl('../oldebvisite')
    }
});

File: remocra/src/main/java/fr/sdis83/remocra/domain/remocra/OldebVisite.java
package fr.sdis83.remocra.domain.remocra;

import java.util.Date;
import java.util.HashSet;
import java.util.Set;

import javax.persistence.CascadeType;
import javax.persistence.Column;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.JoinColumn;
import javax.persistence.JoinTable;
import javax.persistence.ManyToMany;
import javax.persistence.ManyToOne;
import javax.persistence.OneToMany;
import javax.persistence.Temporal;
import javax.persistence.TemporalType;
import javax.validation.constraints.NotNull;

import org.hibernate.annotations.Formula;
import org.hibernate.annotations.OnDelete;
import org.hibernate.annotations.OnDeleteAction;
import org.springframework.format.annotation.DateTimeFormat;
import org.springframework.roo.addon.javabean.RooJavaBean;
import org.springframework.roo.addon.jpa.activerecord.RooJpaActiveRecord;
import org.springframework.roo.addon.json.RooJson;
import org.springframework.roo.addon.tostring.RooToString;

import fr.sdis83.remocra.domain.utils.RemocraDateHourTransformer;

@RooJavaBean
@RooJpaActiveRecord(versionField = "", table = "oldeb_visite", schema = "remocra")
@RooJson
@RooToString
public class OldebVisite {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id")
    private Long id;

    @NotNull
    private String code;

    @ManyToMany
    @JoinTable(name = "oldeb_visite_anomalie", joinColumns = { @JoinColumn(name = "visite", nullable = false) }, inverseJoinColumns = {
            @JoinColumn(name = "anomalie", nullable = false) })
    private Set<TypeOldebAnomalie> typeOldebAnomalies;

    @OnDelete(action = OnDeleteAction.CASCADE)
    @OneToMany(cascade = CascadeType.ALL, mappedBy = "visite", orphanRemoval = true)
    private Set<OldebVisiteDocument> oldebVisiteDocuments = new HashSet<OldebVisiteDocument>();

    @OneToMany(mappedBy = "visite")
    private Set<OldebVisiteSuite> oldebVisiteSuites;

    @ManyToOne
    @JoinColumn(name = "oldeb", referencedColumnName = "id", nullable = false)
    private Oldeb oldeb;

    @ManyToOne
    @JoinColumn(name = "action", referencedColumnName = "id", nullable = false)
    private TypeOldebAction action;

    @ManyToOne
    @JoinColumn(name = "avis", referencedColumnName = "id", nullable = false)
    private TypeOldebAvis avis;

    @ManyToOne
    @JoinColumn(name = "debroussaillement_parcelle", referencedColumnName = "id", nullable = false)
    private TypeOldebDebroussaillement debroussaillementParcelle;

    @ManyToOne
    @JoinColumn(name = "debroussaillement_acces", referencedColumnName = "id", nullable = false)
    private TypeOldebDebroussaillement debroussaillementAcces;

    @ManyToOne
    @JoinColumn(name = "utilisateur", referencedColumnName = "id")
    private Utilisateur utilisateur;

    @Column(name = "date_visite")
    @Temporal(TemporalType.TIMESTAMP)
    @DateTimeFormat(pattern = RemocraDateHourTransformer.FORMAT)
    private Date dateVisite;

    @Column(name = "agent")
    @NotNull
    private String agent;

    @Column(name = "observation")
    private String observation;

    @Formula("(select av.nom from remocra.type_oldeb_avis av where av.id = avis)")
    private String nomAvis;

    @Formula("(select ac.nom from remocra.type_oldeb_action ac where ac.id = action)")
    private String nomAction;

    @Formula("(select tod.nom from remocra.type_oldeb_debroussaillement tod where tod.id = debroussaillement_acces)")
    private String nomDebAcces;

    @Formula("(select tod.nom from remocra.type_oldeb_debroussaillement tod where tod.id = debroussaillement_parcelle)")
    private String nomDebParcelle;

    @Formula("(select count(*) from remocra.oldeb_visite_anomalie ova where ova.visite = id)")
    private String totalAnomalies;
}
